source/agora/common/BlockStorage.d

```diff
@@ -118,37 +234,60 @@ public class BlockStorage
 
     ***************************************************************************/
 
-    public bool saveBlock (const ref Block block)
+    public bool saveBlock (const ref Block block) @safe
     {
-        if (block.header.height != this.block_positions.length)
+        if ((this.height_idx.length > 0) &&
+            (this.height_idx.back.height >= block.header.height))
             return false;
 
         const ubyte[] serialized_block = serializeFull(block);
```

Comment:
Instead of allocating memory to serialize, it should be done straight to the file.